accepted/2025/rid-specific-tool-packages.md

```diff
@@ -0,0 +1,91 @@
+
+## Summary
+
+This proposal adds the ability for a .NET Tool to have separate packages for each supported OS and architecture.
+
+In .NET, the combination of OS and architecture is represented by a Runtime Identifier (RID).  Today, .NET Tools support native assets, but the native assets for all of the supported Runtime Identifiers need to be included in the same package.  For tools with large native dependencies, or if the entire tool is native (for example via NativeAOT), this multiplies the package download size by the number of supported RIDs.
+
+We would like to add support for RID-specific .NET Tool packages.  This would enable a tool to have a separate package for each supported RID, and when the tool is installed only the package for the correct RID would need to be downloaded.
+
+### Goals
+
+- Tools can create RID-specific packages
+- Installing a tool with RID-specific packages is transparent but only downloads the package for the current RID
+
+### Non-goals
+
+- Allowing a tool to have assets from a mixture of packages, for example have architecture neutral assets in the primary package and native dependencies in RID-specific packages
+
+## Tool manifests
+
+.NET Tools have a tool manifest file named `DotnetToolSettings.xml`.  It is stored alongside the executable tool assets in the `tools\<TargetFramework>\<RuntimeIdentifier>` folder.  Currently for most or all tools, the `RuntimeIdentifier` is `any`.  Here is an example of a current manifest:
+
+```xml
+<DotNetCliTool Version="1">
+  <Commands>
+    <Command Name="dotnet-say" EntryPoint="dotnet-say.dll" Runner="dotnet" />
+  </Commands>
+</DotNetCliTool>
+```
+
+## Design
+
+A tool with RID-specific packages will consist of a single primary package and RID-specific packages for each supported RID.  The primary package will include a tool manifest that lists the RIDs supported by the tool, and the package name and package version of the RID-specific tool package for each one.  The primary package will not include any tool implementation assets or shims.  The RID-specific packages will have the same layout and format as tool packages currently have.  The only difference will be that the NuGet package type will be set to a new `DotnetToolRidPackage` type, in order to prevent tool search results from being cluttered with RID-specific tool packages (the primary package is the one that should show up in the results).
+
+The RID-specific tool packages should be named using the convention `<Toolname>.<RID>`.  For example, if the `dotnetsay` tool has RID-specific packages, they would be named `dotnetsay.win-x64`, `dotnetsay.linux-x64`, etc.
+
+The tool manifest for the primary package would look like this:
+
+```xml
+<DotNetCliTool Version="1">
+  <Commands>
+    <Command Name="dotnet-say" />
+  </Commands>
+  <RuntimeIdentifierPackages>
+    <RuntimeIdentifierPackage RuntimeIdentifier="win-x64" Id="dotnet-say.win-x64" Version="1.0.0" />
```

Comment:
This feels a lot like the dependencies section of .nuspec. Akin to @jkotas's [comment above](https://github.com/dotnet/designs/pull/333/files#r2051960057), could this concern be considered more generally and expressed in the .nuspec as a RID-specific dependency?

Reply:
There is the `runtime.json` file in some packages... but I forget how this works...

Reply:
I think we can close this thread in favor of https://github.com/dotnet/designs/pull/333#discussion_r2051960057.